---
title: "Project2"
output: html_notebook
---

## Introduction
 
**Scientific Question**: How accurate is the Tyrer Cuzick risk assessment score that classify people at either average or high risk for breast cancer? 

**Background**: The Tyrer-Cuzick Risk Assessment Calculator is a risk calculator that classifies patients based on their possibility of developing breast cancer. The calculator asks many questions based on the patient beginning with the age, height and weight, age of menstrual period, etc. The assessment also asks about the family history such as whether a parent or both have had cancer. Also, the assessment asks whether the patient had been tested for BRCA1/BRCA2 mutation. Based on these questions, the assessment outputs the possibility of the patient developing breast cancer over the next 10 years.  

**Scientific Hypothesis**: If patients that are classified to be at high-risk of cancer show differently expressed genes, then those genes will help explain the possible breast cancer signatures for further research.  

Here we perform a pairwise sequence alignment of the most common mutations of breast cancer: BRCA1 and BRCA2. Then, we perform differential expression analysis and use various visualizations to see how genes are expressed in people at high vs. average risk of cancer. The purpose is to be able to find any cancer causing signatures that might allow us to further quantify for potential risk of cancer. The file was downloaded from the Gene Expression Omnibus from the NCBI website with an accesion number of GSE164641 from the following paper (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8830042/). 

## Package Descriptions

The packages needed are as follows with a brief description of them: 

**1. Biostrings**: This package is used to analyze biological sequences such as pairwise sequence alignment. (Source: https://bioconductor.org/packages/release/bioc/html/Biostrings.html)

**2. seqinr**: This package is used to perform analysis on DNA or proteins sequence such as reading in the fasta file. (Source: https://rdrr.io/rforge/seqinr/)

**3. DESeq2**: This package is used to perform differential gene expression analysis with an input of a countdata as a matrix. (Source: https://bioconductor.org/packages/devel/bioc/manuals/DESeq2/man/DESeq2.pdf)

**4. RColorBrewer**: This package is used to create visualizations with many colors. (Source: https://cran.r-project.org/web/packages/RColorBrewer/index.html)

**5. pheatmap**: This package is used to create heatmaps after performing differential expression analysis for visualization. (Source: https://cran.r-project.org/web/packages/pheatmap/index.html)

**6. tidyverse**: The tidyverse package include many other packages that are useful for data analyses. Here, we use the readxl package within the tidyverse package to read in the excel file for anlaysis. (Source: https://www.tidyverse.org/packages/)

**7.GEOquery**: This package allows us to perform analysis on GEO datasets such as unzipping the files to create a raw countData. 

**8.readxl**: This package allows us to analyze the excel file by reading in the excel file to create a dataframe.


## Bioinformatics Analysis 1: Pairwise Sequence Alignment

Pairwise sequence alignment is performed to identify how the two sequences of BRCA1 and BRCA2 are similar to each other to see how the sequences indicate the functional relationships between them. 
```{r}
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("Biostrings")
library(Biostrings)
library(seqinr) 

data(BLOSUM50)

# UniProt: BRCA1 (P38398), BRCA2 (P51587)
# The read.fasta function reads in the fasta file (input) 
BRCA1 <- read.fasta("P38398.fasta.txt")
BRCA2 <- read.fasta("P51587.fasta.txt")

# Source: https://community.rstudio.com/t/pairwise-alignment-error-message/17902/7
seq1 <- as.character(BRCA1[[1]])
seq1<-paste(seq1, collapse="")
seq1 <- toupper(seq1)

seq2 <- as.character(BRCA2[[1]])
seq2 <- paste(seq2, collapse="")
seq2 <- toupper(seq2)

# pairwiseAlignment takes in two sequences as input with a scoring matrix, and scores for gapOpening and gapExtension and performse pairwise sequence alignment on the input sequences
globalAligns3s4 <- pairwiseAlignment(seq1, seq2, substitutionMatrix = "BLOSUM50", gapOpening = -2, gapExtension = -8, scoreOnly = FALSE)

globalAligns3s4
```


## Bioinformatics Analysis 2: Differential Expression Analysis

Differential expression analysis is performed by using the count data, which is a matrix of the raw counts for each gene based on the sample. Through differential expression analysis, we seek to find the changes of gene expression between the two groups divided by their risk for breast cancer: High vs. Average. 
```{r}
# Attach the packages needed for analysis 
library(DESeq2)
library(RColorBrewer)
library(pheatmap)
library(tidyverse)
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("GEOquery")
library(GEOquery)
library(readxl)
```

```{r}
# Set my_dir to the directory that has the RAW file downloaded
my_dir <- "/Users/seungmin/Desktop/BIMM143/Project2"

# untar the .tar file 
untar("GSE164641_RAW.tar")

# set the list of files within the RAW.tar file to zip_file
zip_file <- list.files(path=my_dir, pattern="*.gz", full.names=TRUE)

#unzip all files 
lapply(zip_file, gunzip)
```

```{r}
# Create rawcounts data by combining all raw ftcounts.txt 
txt_files <- list.files(path=my_dir, patter="*.ftcounts.txt")

# The lapply function takes in a list of objects and some function and applies the function onto the list of objects
txt_files_df <- lapply(txt_files, function(x) {read.table(file = x, header = TRUE, sep ="\t", row.names="gene")})

# Create a single dataframe "rawcounts" by combining all individual txt files 
# The cbind function combines multiple dataframes into one by columns
rawcounts <- do.call("cbind", lapply(txt_files_df, as.data.frame)) 
rawcounts
```

```{r}
# Create metadata data frame 

# Read in excel file
my_data <- read_excel("13148_2022_1239_MOESM1_ESM.xlsx", sheet=1)

# Create dataframe from the excel file above
my_dataframe <- data.frame(my_data, header=FALSE)

# Set the column names as the first row of dataframe
colnames(my_dataframe) <- my_dataframe[1,]

# Remove first and second and last row for metadata
my_dataframe <- slice(my_dataframe, 3:n()-1)

# Set the row names as the Barcode column
rownames(my_dataframe) <- my_dataframe$Barcode

# Create metadata with rownames and Risk factor
metadata <- subset(my_dataframe, select="Risk")

metadata

# In the rawcounts dataframe, there are donors who have donated their sample twice. Therefore, we have to create a function that gets rid of the second donations from those donors to match the metadata file. 
remove_second_donations <- function(rawcounts, metadata){
  col <- which(!(colnames(rawcounts) %in% rownames(metadata)))
  extract_rawcount <- rawcounts[,-col]
  return (extract_rawcount)
}

extract_rawcounts <- remove_second_donations(rawcounts, metadata)

# There were few mismatches for the Barcodes, so we get rid of the mistmatched ones
row <- which(!(rownames(metadata) %in% colnames(rawcounts)))

extract_metadata <- metadata[-row, ,FALSE]

# Match the order of rownames(extract_metadata) and colnames(extract_rawcount)
reorder <- match(rownames(extract_metadata), colnames(extract_rawcounts))
reordered_rawcounts <- extract_rawcounts[ , reorder]

# Check again that the order of rownames and colnames are equal
identical(rownames(extract_metadata), colnames(reordered_rawcounts))
```

```{r}
# Source: https://app.datacamp.com/learn/courses/rna-seq-with-bioconductor-in-r
# Create a DESeq2 Object 
dds <- DESeqDataSetFromMatrix(countData = reordered_rawcounts, colData = extract_metadata, design = ~ Risk)

# Normalize counts with the DESeq2 Object created
dds <- estimateSizeFactors(dds)
dds_normalized_counts <- counts(dds, normalized=TRUE)
```

```{r}
# Perform DESeq
dds <- DESeq(dds)
res <- results(dds, contrast=c("Risk", "High", "Average"), alpha=0.05)
summary(res)
```


We will be using a volcano plot and heatmap for visualization. Heatmaps allow us visualize the hierarchical clustering that we have performed, showing how genes are differentially expressed based on the risk factor. Volcano plots allow us visualize the genes expressed by the log2FoldChange value, showing how many genes are up- and down- regualted. 
```{r}
# Generate logical column 
res_all <- data.frame(res) 
res_sig <- subset(res, padj < 0.05)

res_all <- data.frame(res) %>% mutate(threshold = padj < 0.05)

# Create the volcano
ggplot(res_all) + 
        geom_point(aes(x = log2FoldChange, y = -log10(padj), color = threshold)) + 
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") + 
        theme(legend.position = "top", 
              plot.title = element_text(size = rel(1.5), hjust = 0.5), 
              axis.title = element_text(size = rel(1.25)))+ scale_color_manual(values = c("FALSE" = "black", "TRUE" = "green"))
```

```{r}
# Subset normalized counts to significant genes
sig_norm_genes <- dds_normalized_counts[rownames(res_sig), ]

# Choose heatmap color palette
heat_colors <- brewer.pal(n = 6, name = "YlOrRd")

# Plot heatmap
pheatmap(sig_norm_genes, 
         color = heat_colors, 
         cluster_rows = T, 
         show_rownames = F,
         annotation = dplyr::select(extract_metadata, Risk), 
         scale = "row")
```


### Analysis: 

When differential expression analysis was performed on the raw counts, we were able to see some genes highly expressed in the Average risk factor and some in the High risk factor based on the heatmap above. However, there was not a clear distinction in the genes differentially expressed based on the risk factor. Therefore, we can conclude that the risk score used for categorizing people into the average vs. high risk factor is not accurate. 



